ios/safe-area/RNSSafeAreaViewComponentView.mm

```diff
@@ -23,44 +48,162 @@ - (instancetype)initWithFrame:(CGRect)frame
+{
+  [self invalidateSafeAreaInsets];
+}
+
+- (void)invalidateSafeAreaInsets
+{
+  if (_providerView == nil) {
+    return;
+  }
+
+  UIEdgeInsets safeAreaInsets = _providerView.RNS_safeAreaInsets;
 
+  if (UIEdgeInsetsEqualToEdgeInsetsWithThreshold(safeAreaInsets, _currentSafeAreaInsets, 1.0 / RCTScreenScale())) {
+    return;
+  }
+
+  _currentSafeAreaInsets = safeAreaInsets;
 #if RCT_NEW_ARCH_ENABLED
-#pragma mark - RCTViewComponentViewProtocol
+  [self updateState];
+#else
+  [self updateLocalData];
+#endif // RCT_NEW_ARCH_ENABLED
+}
+
+#if RCT_NEW_ARCH_ENABLED
+- (void)updateState
+{
+  using facebook::react::RNSSafeAreaViewShadowNode;
+
+  if (!_state) {
+    return;
+  }
+
+  _state->updateState(
+      [=](RNSSafeAreaViewShadowNode::ConcreteState::Data const &oldData)
+          -> RNSSafeAreaViewShadowNode::ConcreteState::SharedData {
+        auto newData = oldData;
+        newData.insets = RCTEdgeInsetsFromUIEdgeInsets(_currentSafeAreaInsets);
+        return std::make_shared<RNSSafeAreaViewShadowNode::ConcreteState::Data const>(newData);
+      });
+}
+#else
+- (void)updateLocalData
+{
+  if (_providerView == nil) {
+    return;
+  }
+  RNSSafeAreaViewLocalData *localData = [[RNSSafeAreaViewLocalData alloc] initWithInsets:_currentSafeAreaInsets
+                                                                                   edges:_edges];
+  [_bridge.uiManager setLocalData:localData forView:self];
+}
+#endif // RCT_NEW_ARCH_ENABLED
 
-//- (void)updateProps:(const facebook::react::Props::Shared &)props
-//           oldProps:(const facebook::react::Props::Shared &)oldProps
-//{
-//  [super updateProps:props oldProps:oldProps];
-//}
+BOOL UIEdgeInsetsEqualToEdgeInsetsWithThreshold(UIEdgeInsets insets1, UIEdgeInsets insets2, CGFloat threshold)
+{
+  return ABS(insets1.left - insets2.left) <= threshold && ABS(insets1.right - insets2.right) <= threshold &&
+      ABS(insets1.top - insets2.top) <= threshold && ABS(insets1.bottom - insets2.bottom) <= threshold;
+}
 
-//- (void)prepareForRecycle
-//{
-//  [super prepareForRecycle];
-//}
+#if RCT_NEW_ARCH_ENABLED
+#pragma mark - RCTViewComponentViewProtocol
 
 + (react::ComponentDescriptorProvider)componentDescriptorProvider
 {
   return react::concreteComponentDescriptorProvider<react::RNSSafeAreaViewComponentDescriptor>();
 }
 
+- (void)updateState:(facebook::react::State::Shared const &)state
+           oldState:(facebook::react::State::Shared const &)oldState
+{
+  using facebook::react::RNSSafeAreaViewShadowNode;
+  _state = std::static_pointer_cast<RNSSafeAreaViewShadowNode::ConcreteState const>(state);
+}
+
+- (void)finalizeUpdates:(RNComponentViewUpdateMask)updateMask
+{
+  [super finalizeUpdates:updateMask];
+  [self updateStateIfNecessary];
+}
+
+- (void)prepareForRecycle
+{
+  [super prepareForRecycle];
+
+  [NSNotificationCenter.defaultCenter removeObserver:self];
+  _state.reset();
+  _providerView = nil;
+  _currentSafeAreaInsets = UIEdgeInsetsZero;
+}
```

Comment:
What if this component is not recycled? Even if the view recycling is enabled, the component is not guaranteed to be recycled, as the pool of recycled components might be full. Will we leak resources in such scenario?